### ISSUE TITLE
Difficulty Reproducing MVP Validation Loss From Checkpoint

### ISSUE THREAD

**Issue by xanderdunn** [timeseriesAI/tsai#94]
I've got my usual `MVP` training setup, and I added a resume function to restore the model and continue training from an existing checkpoint. 
```python
    batch_transformers = [TSStandardize(by_sample=True)]
    print("Creating dataset and data loaders based on TSMetaDataset...")
    datasets = [TSUnwindowedDataset(X, y=y, y_func=None, window_size=window_size, stride=stride, seq_first=True)]
    metadataset = TSMetaDataset(datasets)
    print(metadataset[0], len(metadataset))
    if not splits:
        splits = TSSplitter(valid_size=validation_percentage, show_plot=False)(metadataset)
        print(f"Train split is indices {splits[0][0]}-{splits[0][-1]}")
        print(f"Validation split is indices {splits[1][0]}-{splits[1][-1]}") # type: ignore
    metadatasets = TSMetaDatasets(metadataset, splits)
    print(f"{len(metadatasets.train):,} train") # type: ignore
    print(f"{len(metadatasets.valid):,} val") # type: ignore
    assert len(metadatasets.train) + len(metadatasets.valid) == len(metadataset) # type: ignore

    dataloaders = TSDataLoaders.from_dsets(metadatasets.train,
                                           metadatasets.valid,
                                           bs=[batch_size, batch_size],
                                           batch_tfms=batch_transformers,
                                           num_workers=num_workers,
                                           shuffle_train=True,
                                           drop_last=True,
                                           pin_memory=False)

    checkpoint_dir = f"./data/TSBERT/{dataset_id}/"
    learn = ts_learner(dataloaders,
                       architecture,
                       cbs=[MVP(target_dir=checkpoint_dir,
                           fname=f'{dataset_id}',
                           dropout=0.1,
                           save_model=False)],
                       loss_func=MSELossFlat())

    # resume
    checkpoint_path = os.path.join(checkpoint_dir, get_highest_filename(checkpoint_dir))
    print(f"Resuming model and optimizer from checkpoint {checkpoint_path}...")
    transfer_weights(learn.model, Path(checkpoint_path), exclude_head=True)
    checkpoint = torch.load(checkpoint_path)
    if learn.opt is None: learn.create_opt()
    learn.opt.load_state_dict(checkpoint["optimizer_state_dict"]) # type: ignore
    epoch = checkpoint["epoch"]
    val_loss = checkpoint["val_loss"]
    print(f"Restored checkpoint from epoch {epoch} with val loss {val_loss}.")
    # Do a validation run and see that it matches checkpoint["val_loss"]
    print(learn.validate())
    time.sleep(1)
    print(learn.validate())
    print(learn.validate())
    print(learn.validate())
    print(learn.validate())
```

Here is the output on one run:
```
esuming model and optimizer from checkpoint ./data/TSBERT/my_data/my_data_2.tar...
weights from data/TSBERT/my_data/my_data_2.tar successfully transferred!

Restored checkpoint from epoch 2 with val loss 0.9588498473167419.
validate 1 0.9483685493469238
validate 2 1.0563063621520996
validate 3 1.3346400260925293
validate 4 1.0984760522842407
validate 5 1.6454429626464844
```

And again a second run:
```
Resuming model and optimizer from checkpoint ./data/TSBERT/my_data/my_data_2.tar...
weights from data/TSBERT/my_data/my_data_2.tar successfully transferred!

Restored checkpoint from epoch 2 with val loss 0.9588498473167419.
validate 1 0.9483685493469238
validate 2 1.0563063621520996
validate 3 1.3346400260925293
validate 4 1.0984760522842407
validate 5 1.6454429626464844
```

Notice that the values between the first and second runs are identical, so this is at least reproducible between runs.
The val loss when the run was saved was 0.9588, whereas the first validation I'm seeing after restoring is actually better at 0.9483. The real problem though is that subsequent calls to `.validate()` produce very different and much worse validation losses, almost double the original score. Do you have any thoughts on what might be causing this inconsistency? Surely the batch standardization is the same between validation runs? fastai's `Learner` class calls `no_grad` on `validate()` [here](https://github.com/fastai/fastai/blob/d28a0277a7443ac96472ff12eda96bebde9abbe8/fastai/learner.py#L194) so the weights shouldn't be changing. I tried calling `set_seed()` between calls to `validate()` but it didn't have much of an effect.

**Comment by xanderdunn** [comment]
Ok, I figured it out. It has to do with the head of the `MVP` model. There are two issues. First, when calling `transfer_weights` I need to set `exclude_head=False` so that the `model.head` weights are also restored. 

Second, the `MVP` `__init__` re-initializes the `model.head`, and it's called by the `Learner` at the start of every independent run. This means the head is reset and the weights wiped clean between each of the `.validate` calls in my above code. Hence the loss is mangled. 

To solve this I modified the `MVP` `__init__` so that it doesn't set `model.head` if it's already been set. Then after the first time the MVP's `__init__` is called I can restore all weights including the head and I see the reproducible validation loss that I'm expecting:
```python
        if not hasattr(model, "head_initialized") or not model.head_initialized:
            print("Initializing the MVP model head:")
            model.head = nn.Sequential(nn.Dropout(self.dropout),
                                       nn.Conv1d(model.head_nf, self.learn.dls.vars, 1)).to(self.learn.dls.device)
            model.head_initialized = True
```
```
Resuming model and optimizer from checkpoint ./data/TSBERT/my_data/my_data_4.tar...
check unmatched_layers: ['head.1.0.weight', 'head.1.0.bias']
Restored checkpoint from epoch 4 with val loss 0.9698664546012878.
creating the model head:
validate 1 1.0193445682525635 # Expect the first one to still be broken
weights from data/TSBERT/my_data/my_data_4.tar successfully transferred!
validate 2 0.9660883545875549 # Now that the entire model including the head is restored, all of these losses are good
validate 3 0.9769517183303833
validate 4 0.9624438881874084
validate 5 0.9545067548751831
```
The validation losses vary within +/- 0.01 of the original 0.969 obtained during the original training run. It might be interesting to understand the source of that variance. Dropout? Batch standardization? But this is at least usable.

**Comment by oguiza** [comment]
Hi @xanderdunn,
I've added a parameter (replace_head) that defaults to True. So in general a new head will be created. But for cases like the one you mention, you can set it to False, and it will use the original model head as long as it's capable of recreating the input shape (3d). 
I've also looked at the difference in validation losses. The reason for the discrepancy is that masks applied are random, so some instances may be easier than others. Bear in mind that in MVP we also apply masks to the validation set. Otherwise, the validation set would always be zero, as there would be nothing to guess.

**Comment by xanderdunn** [comment]
Oh of course, the masks would be a source of randomness in the validation loss. I forgot about that. Thank you

**Comment by xanderdunn** [comment]
@oguiza When do you imagine setting `MVP(replace_head=True)`? As the code works right now, the `MVP` callback will replace the model's head for self-supervised pertaining. To resume the self-supervised pre-training, I would still need it to replace the head so that when the weights are loaded with `transfer_weights(exclude_head=False)`, it is the same head being restored. If `MVP(replace_head=false)` were set, then the wrong head would be in the architecture and restoring the weights to the head won't work.

**Comment by oguiza** [comment]
Thanks for your comment. The process I had in mind was unnecessarily complex. 
I've come up with a simpler approach.
First you start pretraining a model like you would normally do. For example: 
```
# Start pretraining
dsid = 'MoteStrain'
X, y, splits = get_UCR_data(dsid, split_data=False)
tfms  = [None, [Categorize()]]
batch_tfms = [TSStandardize(by_var=True)]
unlabeled_dls = get_ts_dls(X, splits=splits, tfms=tfms, batch_tfms=batch_tfms)
learn = ts_learner(unlabeled_dls, InceptionTimePlus, cbs=[MVP(fname=f'{dsid}')])
learn.fit_one_cycle(1, 3e-3)
```
By default save_best will be set to True. That means that the model state_dict() will be saved after each epoch with a lower valid loss. 
You may interrupt pretraining at any point. 
When you want to resume pretraining, you just need to pass the pretrained weights that were previously used (the best ones): 
```
# Continue pretraining from a checkpoint
learn = ts_learner(unlabeled_dls, InceptionTimePlus, cbs=[MVP(weights_path=f'data/MVP/{dsid}.pth')])
learn.fit_one_cycle(1, 3e-3)
```
And the fine-tune: 
```
# Fine-tune
tfms  = [None, [Categorize()]]
batch_tfms = [TSStandardize(by_var=True)]
labeled_dls = get_ts_dls(X, y, splits=splits, tfms=tfms, batch_tfms=batch_tfms, bs=64)
learn = ts_learner(labeled_dls, InceptionTimePlus, pretrained=True, weights_path=f'data/MVP/{dsid}.pth', metrics=accuracy)
learn.fit_one_cycle(1)
```
Please, let me know if this would work
